File: weekly-contest/weekly-contest-170/2-xor-queries-of-a-subarray/prefix-sum.cpp
/**
 * 前缀XOR：因为XOR同时表示“加”“减”
 */

#include <vector>
using namespace std;

class Solution {
public:
    vector<int> xorQueries(vector<int>& arr, vector<vector<int>>& queries) {
        const int n = arr.size();
        vector<int> prefix = {0};
        for (int a : arr) {
            prefix.push_back(prefix.back() ^ a);
        }

        vector<int> res;
        for (auto query : queries) {
            res.push_back(prefix[query[1] + 1] ^ prefix[query[0]]);
        }
        return res;
    }
};

File: weekly-contest/weekly-contest-225/3/prefix-sum.js
/**
 * 前缀和（容斥原理）
 * 
 * 时间：O(N^2)
 */
var kthLargestValue = function(matrix, k) {
  const [R, C] = [matrix.length, matrix[0].length];
  const prefix = Array.from({ length: R + 1 }, () => Array.from({ length: C + 1 }, () => 0));
  const arr = [];
  for (let r = 1; r <= R; ++r) {
    for (let c = 1; c <= C; ++c) {
      prefix[r][c] = matrix[r - 1][c - 1] ^ prefix[r - 1][c] ^ prefix[r][c - 1]
        ^ prefix[r - 1][c - 1];
      arr.push(prefix[r][c]);
    }
  }
  arr.sort((a, b) => b - a);
  return arr[k - 1];
};